infra/gravity/cluster_status.go

```diff
@@ -2,45 +2,106 @@ package gravity
-	ctx, cancel := context.WithTimeout(c.ctx, c.timeouts.Status)
-	defer cancel()
+// statusValidator returns nil if the Gravity Status is the expected status or an error otherwise.
+type statusValidator func(s GravityStatus) error
 
-	retry := wait.Retryer{
-		Attempts: 100,
-		Delay:    time.Second * 20,
+// checkNotDegraded returns an error if the cluster status is Degraded.
+//
+// This function is a reimplementation of the logic in https://github.com/gravitational/gravity/blob/7.0.0/lib/status/status.go#L180-L185
+func checkNotDegraded(s GravityStatus) error {
+	if s.Cluster.State == constants.ClusterStateDegraded {
+		return trace.CompareFailed("cluster state %q", s.Cluster.State)
 	}
+	if s.Cluster.SystemStatus != constants.SystemStatus_Running {
+		return trace.CompareFailed("expected system_status %v, found %v", constants.SystemStatus_Running, s.Cluster.SystemStatus)
+	}
+	return nil
+}
 
-	err := retry.Do(ctx, func() error {
-		errs := make(chan error, len(nodes))
+// checkActive returns an error if the cluster is degraded or state != active.
+func checkActive(s GravityStatus) error {
+	if err := checkNotDegraded(s); err != nil {
+		return trace.Wrap(err)
+	}
+	if s.Cluster.State != constants.ClusterStateActive {
+		return trace.CompareFailed("expected state %q, found %q", constants.ClusterStateActive, s.Cluster.State)
+	}
+	return nil
+}
 
-		for _, node := range nodes {
-			go func(n Gravity) {
-				_, err := n.Status(ctx)
-				errs <- err
-			}(node)
-		}
+// WaitForActiveStatus blocks until all nodes report state = Active and notDegraded or an internal timeout expires.
+func (c *TestContext) WaitForActiveStatus(nodes []Gravity) error {
+	c.Logger().WithField("nodes", Nodes(nodes)).Info("Waiting for active status.")
+	return c.WaitForStatus(nodes, checkActive)
+}
+
+// WaitForStatus blocks until all nodes satisfy the expected statusValidator or an internal timeout expires.
+func (c *TestContext) WaitForStatus(nodes []Gravity, expected statusValidator) error {
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = c.timeouts.ClusterStatus
 
-		err := utils.CollectErrors(ctx, errs)
-		if err == nil {
-			return nil
+	expectStatus := func() (err error) {
+		statuses, err := c.Status(nodes)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		for _, status := range statuses {
+			err = expected(status)
+			if err != nil {
+				c.Logger().WithError(err).WithField("status", status).Warn("Unexpected Status.")
+				return trace.Wrap(err)
+			}
 		}
-		c.Logger().Warnf("Status not available on some nodes, will retry: %v.", err)
-		return wait.Continue("status not ready on some nodes")
-	})
+		return nil
+	}
+
+	err := wait.RetryWithInterval(c.ctx, b, expectStatus, c.Logger())
 
 	return trace.Wrap(err)
+
+}
+
+// Status queries `gravity status` on each node in nodes.
+func (c *TestContext) Status(nodes []Gravity) ([]GravityStatus, error) {
+	ctx, cancel := context.WithTimeout(c.ctx, c.timeouts.NodeStatus)
```

Comment:
nit: just to clarify - `NodeStatus` configures the status timeout for the whole status operation - not individual nodes?
In case it's per node you might need to offset the total amount by the size of the cluster.

Reply:
Why would running multiple status in parallel take longer timeout than running a single status?  What is a reasonable offset in your opinion?

Reply:
Ah, yes, I did not think about them running in parallel when asking the question - sorry, it's fine.